### ISSUE TITLE
Problems reading grib file

### ISSUE THREAD

**Issue by ClaireBr** [nctoolbox/nctoolbox#61]
I have a grib file with the varaibles: elevation , time, lat and long
while I can read time, lat and long (which are a nx1 matrix), I can'tread the elevation (which is a m x n x p matrix). I used the .data('elevation') method.
Can anybody help me with this issue?

This is the error message:

Error using ncdataset/readdata (line 460)
Java exception occurred:
java.io.FileNotFoundException: data file not found = C:/Windows/system32/mars-web233-20150126144043-13236-22324.grib

	at ucar.nc2.grib.GribCollection.getDataRaf(GribCollection.java:387)

	at ucar.nc2.grib.grib1.Grib1Iosp$DataReader.read(Grib1Iosp.java:675)

	at ucar.nc2.grib.grib1.Grib1Iosp.readDataFromCollection(Grib1Iosp.java:649)

	at ucar.nc2.grib.grib1.Grib1Iosp.readData(Grib1Iosp.java:489)

	at ucar.nc2.NetcdfFile.readData(NetcdfFile.java:1957)

	at ucar.nc2.Variable.reallyRead(Variable.java:841)

	at ucar.nc2.Variable._read(Variable.java:813)

	at ucar.nc2.Variable.read(Variable.java:691)

	at ucar.nc2.dataset.VariableDS.reallyRead(VariableDS.java:519)

	at ucar.nc2.dataset.VariableDS._read(VariableDS.java:503)

	at ucar.nc2.Variable.read(Variable.java:691)

Error in ncdataset/data (line 159)
                    d = obj.readdata(variable);

**Comment by hohonuuli** [comment]
In order to figure out what's going on we need to be able to duplicate what you're seeing. So ...

1. Can you post the troublesome grib file somewhere that we can download it.
2. Can you submit the complete code needed to reproduce the problem.

Thanks

**Comment by ClaireBr** [comment]
Thank you very much.

The code would be very short. 
I justed made a dataset and tried to save the variables.

ds = ncdataset('mars-web232-20150126142506-15942-23079.grib');
ds.variables
ans = 

    'Geopotential_surface'
    'lat'
    'lon'
    'time'

lat = ds.data('lat');    % this works nicely
geo = ds.data('Geopotential_surface'); % this doesn't

but  I can't read the size of the variable:

ds.size('Geopotential_surface')
ans =

           1         256         512

You can download the file here: http://www.filedropper.com/mars-web233-20150126144043-13236-22324

Thanks again

**Comment by acrosby** [comment]
@hohonuuli Did you get this file to download?

**Comment by hohonuuli** [comment]
I did but haven't looked at it yet. I reposted it at [https://drive.google.com/file/d/0B8RHIr7J8XHodG1BV3laWnkxaHc/view?usp=sharing](https://drive.google.com/file/d/0B8RHIr7J8XHodG1BV3laWnkxaHc/view?usp=sharing). I leave this link up for a couple of days.

**Comment by hohonuuli** [comment]
OK, I ran the following and everything worked fine:

```matlab
>> ds = ncdataset('mars-web233-20150126144043-13236-22324.grib');
>> lat = ds.data('lat');
>> geo = ds.data('Geopotential_surface');
>> whos
  Name        Size                  Bytes  Class        Attributes

  ds          1x1                     112  ncdataset              
  geo         1x256x512            524288  single                 
  lat       256x1                    1024  single                 

>> ds.size('Geopotential_surface')

ans =

           1         256         512
```

The data matches the CDL:

```
>> ds.netcdf
 
ans =
 
netcdf mars-web233-20150126144043-13236-22324.grib {
 dimensions:
   lon = 512;
   lat = 256;
   time = 1;
 variables:
   float Geopotential_surface(time=1, lat=256, lon=512);
     :long_name = "Geopotential @ Ground or water surface";
     :units = "m2.s-2";
     :missing_value = NaNf; // float
     :grid_mapping = "GaussianLatLon_Projection";
     :Grib_Variable_Id = "VAR_98-0-128-129_L1";
     :Grib1_Center = 98; // int
     :Grib1_Subcenter = 0; // int
     :Grib1_TableVersion = 128; // int
     :Grib1_Parameter = 129; // int
     :Grib1_Parameter_Name = "Z";
     :Grib1_Level_Type = 1; // int
   float lat(lat=256);
     :units = "degrees_north";
     :_CoordinateAxisType = "Lat";
   float lon(lon=512);
     :units = "degrees_east";
     :_CoordinateAxisType = "Lon";
   int time(time=1);
     :units = "Hour since 1989-01-01T12:00:00Z";
     :standard_name = "time";
     :long_name = "Uninitialized analysis / image product / forecast product valid for RT + P1";
     :_CoordinateAxisType = "Time";

 :Originating_or_generating_Center = "European Centre for Medium Range Weather Forecasts (ECMWF) (RSMC)";
 :Originating_or_generating_Subcenter = "0";
 :Conventions = "CF-1.6";
 :history = "Read using CDM IOSP Grib1Collection";
 :featureType = "GRID";
 :file_format = "GRIB-1";
 :_CoordSysBuilder = "ucar.nc2.dataset.conv.CF1Convention";
}
```
So ... I'm not clear on what problem you are actually having? Can you clarify it for me? Thanks

**Comment by ClaireBr** [comment]
When I try to do the same, I get an error message for "geo = .....", but not for "lat = ..."
The error says:

"Error using ncdataset/readdata (line 460)
Java exception occurred:
java.io.FileNotFoundException: data file not found = C:/Windows/system32/mars-web233-20150126144043-13236-22324.grib"
.... (see first post)

I checked my jave version, I think it should be alright:
ans =
Java 1.7.0_11-b21 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode

I have no idea what causes the error. I must be related to some java issue?
Thanks

**Comment by ClaireBr** [comment]
It works now! I downloaded the older version 20130305 and it works just fine. First I had the version 1.10 (latest). 

Thanks anyway!

**Comment by rsignell-usgs** [comment]
@hohonuuli , this is strange.    I verified the problem @ClaireBr is having with the latest 1.1 version.   

When I cd to the `c:\users\rsignell\downloads` directory with the grib file and try open it using ncdataset, a `<gribfile>.gbx9` file is written into that directory, and a `<gribfile>.ncx` file is written into the matlab home directory where I ran `setup_nctoolbox` from (in this case `c:\users\rsignell\documents\matlab`).   Then when we try to read data, it apparently uses the .ncx file and says it can't find the grib file in the matlab home directory.   Which of course it can't, because the grib file is not there, it's in the `c:\users\rsignell\downloads` directory.

Does this give you any clues?

**Comment by acrosby** [comment]
What version of netcdf-java is everyone using now?

**Comment by ClaireBr** [comment]
My MATLAB java version is: Java 1.7.0_11-b21 with Oracle Corporation Java HotSpot(TM) 64-Bit Server VM mixed mode

I don't know what you mean by netcdf-java. Sorry!

**Comment by rsignell-usgs** [comment]
@acrosby , yeah, it's the jar files.   I just tried replacing the 1.1 java directory (using 4.23) with the 20130305 one (using 4.15) and  then 1.1 works fine.    

This makes me sad for two reasons:
* something breaking between minor releases of same netcdf-java 4.3 code
* I meant for 1.1 to be using the same java files as 20130305.   But somehow I got the new ones.   I wonder if I got other "new" stuff as well by mistake.

**Comment by acrosby** [comment]
@rsignell-usgs  It looks to me that the grib and netcdf versions in the tagged 20130305 are the same as 1.1 based on my github sluthing. Everything at 4.3.15--what should they be?

**Comment by rsignell-usgs** [comment]
They should be at 4.3.15, but when I download the v1.1 release zip file from the [nctoolbox releases page](https://github.com/nctoolbox/nctoolbox/releases) I get the 4.3.23 ones.

**Comment by acrosby** [comment]
Seems like something weird happened to the tagged version. I don't see any commits in master that would have popped 4.3.23 in there.

**Comment by rsignell-usgs** [comment]
The commits were to the 1.x branch.

**Comment by hohonuuli** [comment]
The jars were changed by me on the _develop_ branch in change b516aec. You can revert them if needed. However, my suggestion is that the master branch should contain the actual releases; releases should be tags on the master branch. I don't monkey with the master branch at all; but the develop branch is where I do most of my development. From our repo's git history:

```
* 1ea5804 - 2015-01-20 - (origin/develop, develop) Added an SBT project for testing ad hoc NetCDF commands outside of Matlab. This helps resolve classpath issues that may be occurring due to Matlab's numerous unversioned Java jars <Brian Schlining>
* 354fc6c - 2015-01-20 - keeping copies of geodemos, which will be written to use geocdm insteac of ncgeodataset <Brian Schlining>
* f2150db - 2015-01-20 - check point. <Brian Schlining>
* 31237df - 2014-10-23 - Begin adapting methods from ncgeovariable to geocdmvariable. Removing lots and lots of code duplication <Brian Schlining>
* bc918cb - 2014-10-23 - implemented geocdm. Ready to convert all the supporting functions to use it. <Brian Schlining>
* b516aec - 2014-10-23 - Reverted to NetCDF-java v4.3.23 to address issue #52 <Brian Schlining>
```

**Comment by rsignell-usgs** [comment]
I guess I was thinking that if we are going to maintain a specific version with bugfixes, like 1.0, then we need a 1.x branch, and need to release off that branch, don't we?

Just using master was a problem because master moved on from 20130305 version, but starting having issues.   So I wanted to release something based on the 20130305 version, but with bug fixes.

**Comment by hohonuuli** [comment]
It's good thought. There's a couple of strategies for maintaining versions/bugfixes though. My preferred would be to tag the master branch with releases. If we have a bug fix, it's most likely just going to be the next change set on the master branch, which we can then tag with the new release name. If we have older versions that we decide to maintain and apply patches to, we can just checkout the tag and create a branch from it. That way we don't proliferate tons of branches that don't really contain changes.

**Comment by acrosby** [comment]
Agree. Its not really a complex enough project (like an OS) to need to maintain bugs on completely separate versions I think.
On Wed, Jan 28, 2015 at 5:40 PM, Brian Schlining <notifications@github.com> wrote:It's good thought. There's a couple of strategies for maintaining versions/bugfixes though. My preferred would be to tag the master branch with releases. If we have a bug fix, it's most likely just going to be the next change set on the master branch, which we can then tag with the new release name. If we have older versions that we decide to maintain and apply patches to, we can just checkout the tag and create a branch from it. That way we don't proliferate tons of branches that don't really contain changes.

**Comment by rsignell-usgs** [comment]
Fixed by PR #62
